File: src/main/kotlin/ar/edu/utn/frba/tacs/tp/api/herocardsgame/controllers/DecksController.java
package ar.edu.utn.frba.tacs.tp.api.herocardsgame.controllers;

import ar.edu.utn.frba.tacs.tp.api.herocardsgame.models.game.Card;
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.models.game.Deck;
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.request.AddCardToDeckRequest;
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.request.UpdateNameDeckRequest;
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.utils.FileConstructorUtils;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * El administrador tiene la capacidad de manejar los mazos (crear, eliminar, modificar)
 * y ponerles un nombre
 */
@Controller
public class DecksController {

  @GetMapping("/users/decks")
  public ResponseEntity<List<Deck>> getDecks() {
    Card batman = FileConstructorUtils.createFromFile("src/main/resources/json/card/Batman.json", Card.class);
    Card flash = FileConstructorUtils.createFromFile("src/main/resources/json/card/Flash.json", Card.class);

    Deck batmanDeck = new Deck(1L, "Batman deck", Collections.singletonList(batman));
    Deck flashDeck = new Deck(2L, "Justice League deck", Arrays.asList(batman, flash));

    return ResponseEntity.status(HttpStatus.OK).body(Arrays.asList(batmanDeck, flashDeck));
  }

  /**
   *  TODO we should validate all the cards added has attributes needed for the game.
   * @param deck
   * @return
   */
  @PostMapping("/admin/decks")
  public ResponseEntity<Deck> createDeck(@RequestBody Deck deck) {
    return ResponseEntity.status(HttpStatus.CREATED).body(deck);
  }

  /**
   * TODO use the same validation to create the deck
   * @param deckId
   * @return
   */
  @PutMapping("/admin/decks/{deckId}")
  public ResponseEntity updateDeck(@PathVariable("deckId") String deckId, @RequestBody Deck deck) {
    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
  }

  @PatchMapping("/admin/decks/{deckId}/name")
  public ResponseEntity updateNameDeck(@PathVariable("deckId") String deckId,
                                       @RequestBody UpdateNameDeckRequest updateNameDeckRequest){
    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
  }

  //TODO: Tengo dudas con respecto a si es put o patch
  //TODO: Tengo dudas con respecto a si conviene agregar la carta por id o por objeto
  @PatchMapping("/admin/decks/{deckId}/card")
  public ResponseEntity addCardToDeck(@PathVariable("deckId") String deckId,
                                      @RequestBody AddCardToDeckRequest addCardToDeckRequest){
    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
  }

  @DeleteMapping("/admin/decks/{deckId}/card/{cardId}")
  public ResponseEntity deleteCardToDeck(@PathVariable("deckId") String deckId, @PathVariable("cardId") String cardId){
    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
  }

  @DeleteMapping("/admin/decks/{deckId}")
  public ResponseEntity deleteDeck(@PathVariable("deckId") String deckId) {
    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
  }

}

File: src/main/kotlin/ar/edu/utn/frba/tacs/tp/api/herocardsgame/controllers/DecksController.kt
package ar.edu.utn.frba.tacs.tp.api.herocardsgame.controllers;

import ar.edu.utn.frba.tacs.tp.api.herocardsgame.exception.ElementNotFoundException
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.exception.InvalidPowerstatsException
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.models.game.deck.Deck
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.request.CreateDeckRequest
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.request.UpdateDeckRequest
import ar.edu.utn.frba.tacs.tp.api.herocardsgame.service.DeckService
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*

/**
 * El administrador tiene la capacidad de manejar los mazos (crear, eliminar, modificar)
 * y ponerles un nombre
 */
@Controller
@CrossOrigin(origins = ["http://localhost:5000"], allowedHeaders = ["*"])
class DecksController(private val deckService: DeckService) :
    AbstractController<DecksController>(DecksController::class.java) {

    /**
     * @return list of deck
     */
    @GetMapping("/decks")
    fun getDecks(): ResponseEntity<List<Deck>> {
        reportRequest(
            method = RequestMethod.GET,
            path = "/decks",
            body = null
        )
        val response = deckService.searchDeck()
        return reportResponse(HttpStatus.OK, response)
    }

    /**
     * @param deckId, deckName
     * @return list of deck
     */
    @GetMapping("/decks/search")
    fun getDeckByIdOrName(
        @RequestParam(value = "deck-id") deckId: String?,
        @RequestParam(value = "deck-name") deckName: String?
    ): ResponseEntity<List<Deck>> {
        reportRequest(
            method = RequestMethod.GET,
            path = "/decks/search",
            body = null,
            requestParams = hashMapOf("deck-id" to deckId, "deck-name" to deckName)
        )
        val response = deckService.searchDeck(deckId, deckName)
        return reportResponse(HttpStatus.OK, response)
    }

    /**
     * @param createDeckRequest
     * @return deck
     */
    @PostMapping("/admin/decks")
    fun createDeck(@RequestBody createDeckRequest: CreateDeckRequest): ResponseEntity<Deck> =
        try {
            reportRequest(
                method = RequestMethod.POST,
                path = "/admin/decks",
                body = createDeckRequest
            )
            val response = deckService.createDeck(createDeckRequest.deckName, createDeckRequest.cardIds)
            reportResponse(HttpStatus.CREATED, response)
        } catch (e: ElementNotFoundException) {
            reportError(e, HttpStatus.BAD_REQUEST)
        }catch (e: InvalidPowerstatsException) {
            reportError(e, HttpStatus.BAD_REQUEST)
        }

    /**
     * @param deckId, updateDeckRequest
     * @return
     */
    @PutMapping("/admin/decks/{deck-id}")
    fun updateDeck(
        @PathVariable("deck-id") deckId: String, @RequestBody updateDeckRequest: UpdateDeckRequest
    ): ResponseEntity<Deck> =
        try {
            reportRequest(
                method = RequestMethod.PUT,
                path = "/admin/decks/{deck-id}",
                pathVariables = hashMapOf("deck-id" to deckId),
                body = updateDeckRequest
            )
            val response = deckService.updateDeck(
                deckId,
                updateDeckRequest.deckName,
                updateDeckRequest.deckCards ?: emptyList()
            )
            reportResponse(HttpStatus.OK, response)
        } catch (e: ElementNotFoundException) {
            reportError(e, HttpStatus.BAD_REQUEST)
        }catch (e: InvalidPowerstatsException) {
            reportError(e, HttpStatus.BAD_REQUEST)
        }

    /**
     * @param deckId
     * @return
     */
    @DeleteMapping("/admin/decks/{deck-id}")
    fun deleteDeck(@PathVariable("deck-id") deckId: String): ResponseEntity<Void> =
        try {
            reportRequest(
                method = RequestMethod.DELETE,
                path = "/admin/decks/{deck-id}",
                pathVariables = hashMapOf("deck-id" to deckId),
                body = null
            )
            deckService.deleteDeck(deckId)
            reportResponse(HttpStatus.NO_CONTENT)
        } catch (e: ElementNotFoundException) {
            reportError(e, HttpStatus.BAD_REQUEST)
        }

}